src/metabase/api/common/openapi.clj

```diff
@@ -3,90 +3,75 @@
-    (assoc param
-           :schema (-> (dissoc schema :enum) (assoc :type "boolean"))
-           :required false)
-
-    :else
-    param))
-
-(let [file-schema (json-schema-transform ms/File)]
-  ;; TODO: unify this with `fix-type` somehow, but `:required` is making this hard
-  (defn- fix-schema
-    "Change type of request body to make it more understandable to Rapidoc."
-    [{:keys [required] :as schema}]
-    (let [not-required (atom #{})]
-      (-> schema
-          (update :properties (fn [props]
-                                (into {}
-                                      (for [[k v] props]
-                                        [k
-                                         (cond
-                                           (and (:oneOf v)
-                                                (= (second (:oneOf v)) {:type "null"}))
-                                           (do
-                                             (swap! not-required conj k)
-                                             (merge (first (:oneOf v))
-                                                    (select-keys v [:description :default])))
-
-                                           (= (select-keys v (keys file-schema)) file-schema)
-                                           ;; I got this from StackOverflow and docs are not in agreement, but RapiDoc
-                                           ;; shows file input, so... :)
-                                           (merge {:type "string" :format "binary"}
-                                                  (select-keys v [:description :default]))
-
-                                           (= (:enum v) ["true" "false" true false])
-                                           (-> (dissoc v :enum) (assoc :type "boolean"))
-
-                                           :else
-                                           v)]))))
-          (assoc :required (vec (remove @not-required required)))))))
+      (swap! *definitions* merge (:definitions jss)))
+    (dissoc jss :definitions)))
+
+(defn- merge-required [schema]
+  (let [optional? (set (keep (fn [[k v]] (when (:optional v) k))
+                             (:properties schema)))]
+    (-> schema
+        (m/update-existing :required #(vec (remove optional? %)))
+        (m/update-existing :properties #(update-vals % (fn [v] (dissoc v :optional)))))))
+
+(let [file-schema (mjs/transform ms/File)]
+  (defn- fix-json-schema
+    "Clean-up JSON schema to make it more understandable for OpenAPI tools.
+
+    Returns a new schema WITH an indicator if it's *NOT* required.
+
+    NOTE: maybe instead of fixing it up later we should re-work Malli's json-schema transformation into a way we want it to be?"
+    [schema]
+    (cond
+      ;; we're using `[:maybe ...]` a lot, and it generates `{:oneOf [... {:type "null"}]}`
+      ;; it needs to be cleaned up to be presented in OpenAPI viewers
+      (and (:oneOf schema)
```

Comment:
Should be not check that there aren't more than 2 arguments to `:oneOf`?

Reply:
Maybe for completeness, but in reality we don't do `:multi` with `nil` inside, so it's not very dangerous.